### PR TITLE
test: Move existing benchmarks to use new bazel build extension for benchmark targets

### LINKED ISSUE #9539: [benchmark] //test/extensions/filters/listener/tls_inspector:tls_inspector_benchmark crashes
*Description*:
The benchmark for tls_inspector does not run successfully.

*Repro steps*:
$ bazel run -c opt //test/extensions/filters/listener/tls_inspector:tls_inspector_benchmark

Running /usr/local/google/home/avd/.cache/bazel/_bazel_avd/1e20bf3ad6fa0170c07bc768f0392847/execroot/envoy/bazel-out/k8-opt/bin/test/extensions/filters/listener/tls_inspector/tls_inspector_benchmark
Run on (72 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x36)
  L1 Instruction 32K (x36)
  L2 Unified 1024K (x36)
  L3 Unified 25344K (x2)
Load Average: 2.67, 20.83, 22.49
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
terminating with uncaught exception of type std::__1::bad_function_call: std::exception
Aborted

*Call Stack*:
Caught Aborted, suspect faulting address 0x14e380003e89a
Backtrace (use tools/stack_decode.py to get line numbers):
Envoy version: 559443c430d0284418f3788bede4938b7f40cf03/1.13.0-dev/Clean/RELEASE/BoringSSL
#0: __restore_rt [0x7f26716303a0]
#1: Envoy::Extensions::ListenerFilters::TlsInspector::BM_TlsInspector() [0x52ca86]
#2: benchmark::internal::BenchmarkInstance::Run() [0xc30193]
#3: benchmark::internal::(anonymous namespace)::RunInThread() [0xc2d933]
#4: benchmark::internal::RunBenchmark() [0xc2c076]
#5: benchmark::RunSpecifiedBenchmarks() [0xc27962]
#6: main [0xbdb80d]
#7: __libc_start_main [0x7f267128252b]

### PR DESCRIPTION
**Pull request by antoniovicente**
Description: Add bazel build extensions for benchmarks targets and tests for benchmark targets that expands to a benchmark binary with an implicit dependency on //test/benchmark:main. The new benchmark main registers a signal handler that prints the failure stack on failure, in addition to doing benchmark initialization, flag parsing and running the selected benchmarks.

Also, fix existing benchmarks that didn't use to complete successfully.
Risk Level: n/a
Testing: adds test targets to provide basic coverage for existing benchmarks.
Docs Changes: n/a
Release Notes: n/a
Fixes #9539, #9540, #9541

### CODE REVIEW

File: bazel/envoy_build_system.bzl [hunk 1]
```diff
@@ -21,6 +21,7 @@ load(
 )
 load(
     ":envoy_test.bzl",
+    _envoy_cc_benchmark_binary_and_test = "envoy_cc_benchmark_binary_and_test",
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Nit: prefer `envoy_cc_benchmark`, this is a bit cleaner visually and implies the same information.

File: bazel/test_for_benchmark_wrapper.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+HEAPCHECK= "$@"
```
Comment (htuch) [inline bazel/test_for_benchmark_wrapper.sh]: Can you add a comment explaining what is going on here? I assume this is related to heap profiling.

File: test/benchmark/main.cc [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// NOLINT(namespace-envoy)
+// This is an Envoy driver for benchmarks.
+
+#ifdef ENVOY_HANDLE_SIGNALS
+#include "common/signal/signal_action.h"
+
```
Comment (htuch) [inline test/benchmark/main.cc]: Nit: gratuitous blank line.

File: test/common/access_log/access_log_formatter_speed_test.cc [hunk 4]
```diff
@@ -61,37 +61,38 @@ BENCHMARK(BM_TypedJsonAccessLogFormatter);
...
-  json_formatter = std::make_unique<Envoy::AccessLog::JsonFormatterImpl>(JsonLogFormat, false);
-  typed_json_formatter = std::make_unique<Envoy::AccessLog::JsonFormatterImpl>(JsonLogFormat, true);
-
-  stream_info = std::make_unique<Envoy::TestStreamInfo>();
-  stream_info->setDownstreamRemoteAddress(
-      std::make_shared<Envoy::Network::Address::Ipv4Instance>("203.0.113.1"));
-  benchmark::Initialize(&argc, argv);
-  if (benchmark::ReportUnrecognizedArguments(argc, argv)) {
-    return 1;
+// Hack to initialize the global variables used by the benchmarks above.
```
Comment (htuch) [inline test/common/access_log/access_log_formatter_speed_test.cc]: FWIW, this extant hack is pretty terrible, I hope we can just get rid of it and not rely on globals.

File: test/common/network/lc_trie_speed_test.cc [hunk 5]
```diff
@@ -97,44 +97,45 @@ BENCHMARK(BM_LcTrieLookupMinimal);
...
-int main(int argc, char** argv) {
-
-  // Random test addresses from RFC 5737 netblocks
-  static const std::string test_addresses[] = {
-      "192.0.2.225",   "198.51.100.55", "198.51.100.105", "192.0.2.150",   "203.0.113.162",
-      "203.0.113.110", "203.0.113.99",  "198.51.100.23",  "198.51.100.24", "203.0.113.12"};
-  for (const auto& address : test_addresses) {
-    addresses.push_back(Envoy::Network::Utility::parseInternetAddress(address));
-  }
+// Hack to initialize the global variables used by the benchmarks above.
```
Comment (htuch) [inline test/common/network/lc_trie_speed_test.cc]: Can't we just have a setup method per-benchmark?

File: test/common/access_log/access_log_formatter_speed_test.cc (hunk 4 above)
Comment (antoniovicente) [reply test/common/access_log/access_log_formatter_speed_test.cc]: I don't know how to get rid of it.  Do you have any suggestions?

File: test/common/network/lc_trie_speed_test.cc (hunk 5 above)
Comment (antoniovicente) [reply test/common/network/lc_trie_speed_test.cc]: I think that would require a change to the google benchmarking libraries.  It's doable, but doesn't seem trivial.

File: bazel/envoy_build_system.bzl (hunk 1 above)
Comment (antoniovicente) [reply bazel/envoy_build_system.bzl]: I think I'll split this into 2 build extensions: envoy_cc_benchmark and envoy_benchmark_test.  It does require people adding both to BUILD files, but feels cleaner.

Comment (htuch) [reply bazel/envoy_build_system.bzl]: Up to you, I'm fine with just renaming to `envoy_cc_benchmark` and having multiple targets.

File: test/common/network/lc_trie_speed_test.cc (hunk 5 above)
Comment (htuch) [reply test/common/network/lc_trie_speed_test.cc]: Why can't each benchmark method call into a setup function with the benchmark clock disabled?

Comment (antoniovicente) [reply test/common/network/lc_trie_speed_test.cc]: Refactored how test data is constructed.

**Comment by antoniovicente** [comment]
Seems like some of the benchmarks have some use-after-free and initialization-order-fiasco errors under ASAN.  Give me some time to fix those and address existing comments before you make a pass on the PR.

File: bazel/envoy_build_system.bzl (hunk 1 above)
Comment (antoniovicente) [reply bazel/envoy_build_system.bzl]: Switched to separate benchmark and test rules.

File: bazel/test_for_benchmark_wrapper.sh (hunk 2 above)
Comment (antoniovicente) [reply bazel/test_for_benchmark_wrapper.sh]: Fixed the leak issue, removed HEAPCHECK=

File: test/benchmark/main.cc (hunk 3 above)
Comment (antoniovicente) [reply test/benchmark/main.cc]: fixed.

Comment (antoniovicente) [reply test/benchmark/main.cc]: Fix format is what is adding this newline, it's not me.

File: test/common/access_log/access_log_formatter_speed_test.cc (hunk 4 above)
Comment (antoniovicente) [reply test/common/access_log/access_log_formatter_speed_test.cc]: Fixed.

**Comment by antoniovicente** [comment]
> Seems like some of the benchmarks have some use-after-free and initialization-order-fiasco errors under ASAN. Give me some time to fix those and address existing comments before you make a pass on the PR.

ASAN/TSAN issues fixed.  I think.

**Comment by antoniovicente** [comment]
ping on review.  I don't know what's up with the stale "Changes requested" and never completing CI above, which refers to an old commit.

**Comment by mattklein123** [comment]
cc @tonya11en to help review this while @htuch is digging out. Tony was also interested in seeing this PR happen. Looks like a master merge is needed also.

File: bazel/test_for_benchmark_wrapper.sh [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+# Set the benchmark time to a tiny value since we just want to know if the benchmark runs to completion.
+"$@" --benchmark_min_time=0.00001
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply bazel/test_for_benchmark_wrapper.sh]: done

File: test/benchmark/main.cc [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+// Boilerplate main(), which discovers benchmarks and runs them.
+int main(int argc, char** argv) {
+#ifndef __APPLE__
+  absl::InitializeSymbolizer(argv[0]);
+#endif
+#ifdef ENVOY_HANDLE_SIGNALS
+  // Enabled by default. Control with "bazel --define=signal_trace=disabled"
+  Envoy::SignalAction handle_sigs;
+#endif
+
```
Comment (htuch) [inline test/benchmark/main.cc]: Are we doing enough initialization here to support general purpose Envoy code in benchmarks? E.g. if we want to enable logging while debugging a benchmark, would we want something like https://github.com/envoyproxy/envoy/blob/master/test/test_runner.cc#L108?

File: test/common/buffer/buffer_speed_test.cc [hunk 8]
```diff
@@ -113,8 +113,9 @@ static void BufferPrependBuffer(benchmark::State& state) {
     // buffer every time without the overhead of a copy, we use an BufferFragment that references
     // (and never deletes) an external string.
     Buffer::OwnedImpl to_add;
-    Buffer::BufferFragmentImpl fragment(input.data(), input.size(), DoNotReleaseFragment);
-    to_add.addBufferFragment(fragment);
+    auto fragment =
+        std::make_unique<Buffer::BufferFragmentImpl>(input.data(), input.size(), DeleteFragment);
+    to_add.addBufferFragment(*fragment.release());
```
Comment (htuch) [inline test/common/buffer/buffer_speed_test.cc]: Can you separate out the general "make benchmarks more awesome/readable/correct" from the Bazel aspects, ideally this PR is just Bazel.

File: test/benchmark/main.cc (hunk 7 above)
Comment (antoniovicente) [reply test/benchmark/main.cc]: We need to start from somewhere and iterate.  If you want me to add some log config lines, I can do that.

File: test/common/buffer/buffer_speed_test.cc (hunk 8 above)
Comment (antoniovicente) [reply test/common/buffer/buffer_speed_test.cc]: I can certainly make the benchmark tests blindly first, could I get a fast review on that?

I think the main blocker on this CL right now is the compile_time_options that I can't fully replicate.  There may be some existing breakages in that mode.

Comment (htuch) [reply test/common/buffer/buffer_speed_test.cc]: Yes, happy to provide fast turnaround on PRs here. For compile_time_options, I wonder if `--build_test_only` is interacting somehow in https://github.com/envoyproxy/envoy/blob/master/ci/do_ci.sh#L233?

File: test/benchmark/main.cc (hunk 7 above)
Comment (htuch) [reply test/benchmark/main.cc]: I think making sure logging works would be great.

**Comment by antoniovicente** [comment]
I'm sending smaller PRs to transition individual benchmarks to new build rules.  Closing this PR.

**Pull request closed without merging**